runtime error: invalid memory address or nil pointer dereference

````
INFO [07-25|12:19:10.516] Maximum peer count                       ETH=50 LES=0 total=50
WARN [07-25|12:19:10.516] The flag --rpc is deprecated and will be removed June 2021, please use --http
WARN [07-25|12:19:10.516] The flag --rpcaddr is deprecated and will be removed June 2021, please use --http.addr
WARN [07-25|12:19:10.516] The flag --rpcport is deprecated and will be removed June 2021, please use --http.port
WARN [07-25|12:19:10.516] The flag --rpcapi is deprecated and will be removed June 2021, please use --http.api
WARN [07-25|12:19:10.516] The flag --rpcvhosts is deprecated and will be removed June 2021, please use --http.vhosts
INFO [07-25|12:19:10.516] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [07-25|12:19:10.517] Set global gas cap                       cap=25,000,000
INFO [07-25|12:19:10.527] Configured Ethereum protocol versions    capabilities="[66 65]"
INFO [07-25|12:19:10.527] Allocated trie memory caches             clean=307.00MiB dirty=512.00MiB
INFO [07-25|12:19:10.527] Allocated cache and file handles         database=/home/cgeth/data_multi/geth/chaindata cache=1024.00MiB handles=32767
INFO [07-25|12:19:21.022] Opened ancient database                  database=/home/cgeth/data_multi/geth/chaindata/ancient readonly=false
INFO [07-25|12:19:21.218] Rebuilding state snapshot 
INFO [07-25|12:19:21.218] Resuming state snapshot generation       root=a87d03..8077ee accounts=0 slots=0 storage=0.00B elapsed="255.662µs"
INFO [07-25|12:19:21.218] Loaded local transaction journal         transactions=0 dropped=0
INFO [07-25|12:19:21.219] Regenerated local transaction journal    transactions=0 accounts=0
WARN [07-25|12:19:21.219] Switch sync mode from fast sync to full sync 
WARN [07-25|12:19:21.236] Unclean shutdown detected                booted=2021-07-25T12:17:38+0800 age=1m43s
INFO [07-25|12:19:21.236] Starting peer-to-peer node               instance=CoreGeth/v1.12.0-stable-c5ef3a9e/linux-amd64/go1.16.2
INFO [07-25|12:19:21.364] New local node record                    seq=4 id=5003485a43eba516 ip=127.0.0.1 udp=30306 tcp=30306
INFO [07-25|12:19:21.366] IPC endpoint opened                      url=/home/cgeth/data_multi/etc.ipc
INFO [07-25|12:19:21.366] HTTP server started                      endpoint=[::]:18541 prefix= cors= vhosts=cgeth.bit99.com
INFO [07-25|12:19:21.366] Started P2P networking                   self=enode://5837292b7b3d8d478f1077215f1fbc2f05afda6aa11d6babd2694811f50bba922db1f6d18972714c94e1af9c6e3bdf7a959d7fcd26542fdf3c71b05f34cbbaec@127.0.0.1:30306
INFO [07-25|12:19:29.218] Generating state snapshot                root=a87d03..8077ee at=007d2c..c8652d accounts=145,968 slots=5327 storage=10.14MiB elapsed=8.000s      eta=1h9m41.035s
INFO [07-25|12:19:31.367] Block synchronisation started 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1c0 pc=0xce9cb0]

goroutine 45587 [running]:
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).findAncestorBinarySearch(0xc00693e000, 0xc0169573b0, 0x0, 0xc95de2, 0xb9558a, 0xb9558a, 0x0, 0x1802960)
        /home/ia/go/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:1002 +0x590
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).findAncestor(0xc00693e000, 0xc0169573b0, 0xc000014240, 0x0, 0x0, 0x0)
        /home/ia/go/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:848 +0x3a5
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).syncWithPeer(0xc00693e000, 0xc0169573b0, 0x47796ee80af77e91, 0x5fdbe24c4919b311, 0x9353780fc9cdc27a, 0x9bdcbca791a11ea5, 0xc03700a600, 0x0, 0x0)
        /home/ia/go/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:501 +0xfaf
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).synchronise(0xc00693e000, 0xc00072adc0, 0x40, 0x47796ee80af77e91, 0x5fdbe24c4919b311, 0x9353780fc9cdc27a, 0x9bdcbca791a11ea5, 0xc03700a600, 0x0, 0x0, ...)
        /home/ia/go/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:453 +0x3b1
github.com/ethereum/go-ethereum/eth/downloader.(*Downloader).Synchronise(0xc00693e000, 0xc00072adc0, 0x40, 0x47796ee80af77e91, 0x5fdbe24c4919b311, 0x9353780fc9cdc27a, 0x9bdcbca791a11ea5, 0xc03700a600, 0xc000000000, 0xffa66e, ...)
        /home/ia/go/src/github.com/ethereum/go-ethereum/eth/downloader/downloader.go:348 +0x8c
github.com/ethereum/go-ethereum/eth.(*handler).doSync(0xc006282d00, 0xc0385d4700, 0x16a3ba0, 0xc0044b01c0)
        /home/ia/go/src/github.com/ethereum/go-ethereum/eth/sync.go:383 +0x125
github.com/ethereum/go-ethereum/eth.(*chainSyncer).startSync.func1(0xc000562ae0, 0xc0385d4700)
        /home/ia/go/src/github.com/ethereum/go-ethereum/eth/sync.go:359 +0x38
created by github.com/ethereum/go-ethereum/eth.(*chainSyncer).startSync
        /home/ia/go/src/github.com/ethereum/go-ethereum/eth/sync.go:359 +0x76
````

--- a/eth/downloader/downloader.go
+++ b/eth/downloader/downloader.go
@@ -958,30 +958,31 @@ func (d *Downloader) findAncestorBinarySearch(p *peerConnection, mode SyncMode, 
 				arrived = true
 
 				// Modify the search interval based on the response
-				hash = headers[0].Hash()
+				h := headers[0].Hash()
 				n := headers[0].Number.Uint64()
 
 				var known bool
 				switch SyncMode(d.mode) {
 				case FullSync:
-					known = d.blockchain.HasBlock(hash, n)
+					known = d.blockchain.HasBlock(h, n)
 				case FastSync:
-					known = d.blockchain.HasFastBlock(hash, n)
+					known = d.blockchain.HasFastBlock(h, n)
 				case LightSync:
-					known = d.lightchain.HasHeader(hash, n)
+					known = d.lightchain.HasHeader(h, n)
 				default:
 					log.Crit("unknown sync mode", "mode", d.mode)
 				}
 				if !known {
 					end = check
-					break
+					continue
 				}
-				header := d.lightchain.GetHeaderByHash(hash) // Independent of sync mode, header surely exists
+				header := d.lightchain.GetHeaderByHash(h) // Independent of sync mode, header surely exists
 				if header.Number.Uint64() != check {
 					p.log.Warn("Received non requested header", "number", header.Number, "hash", header.Hash(), "request", check)
 					return 0, fmt.Errorf("%w: non-requested header (%d)", errBadPeer, header.Number)
 				}
 				start = check
+				hash = h
 
 			case <-timeout:
 				p.log.Debug("Waiting for search header timed out", "elapsed", ttl)

